contracts/chain-adapters/Arbitrum_CustomGasToken_L2_Forwarder.sol

```diff
@@ -0,0 +1,184 @@
+
+/**
+ * @notice Interface for funder contract that this contract pulls from to pay for relayMessage()/relayTokens()
+ * fees using a custom gas token.
+ */
+interface FunderInterface {
+    /**
+     * @notice Withdraws amount of token from funder contract to the caller.
+     * @dev Can only be called by owner of Funder contract, which therefore must be
+     * this contract.
+     * @param token Token to withdraw.
+     * @param amount Amount to withdraw.
+     */
+    function withdraw(IERC20 token, uint256 amount) external;
+}
+
+/**
+ * @notice Contract containing logic to send messages from L1 to Arbitrum.
+ * @dev This contract is very similar to Arbitrum_CustomGasToken_Adapter. It is meant to bridge
+ * tokens and send messages over a bridge which uses a custom gas token, except this contract makes
+ * no assumptions about the L2 on which it is deployed.
+ */
+
+// solhint-disable-next-line contract-name-camelcase
+contract Arbitrum_CustomGasToken_L2_Forwarder is ArbitrumForwarderInterface, CircleCCTPAdapter {
+    using SafeERC20 for IERC20;
+
+    // This token is used to pay for l2 to l3 messages if its configured by an Arbitrum orbit chain.
+    IERC20 public immutable CUSTOM_GAS_TOKEN;
+
+    // Contract that funds Inbox cross chain messages with the custom gas token.
+    FunderInterface public immutable CUSTOM_GAS_TOKEN_FUNDER;
+
+    error InvalidCustomGasToken();
+    error InsufficientCustomGasToken();
+
+    modifier onlyFromCrossDomainAdmin() {
+        require(msg.sender == _applyL1ToL2Alias(CROSS_DOMAIN_ADMIN), "ONLY_CROSS_DOMAIN_ADMIN");
+        _;
+    }
+
+    /**
+     * @notice Constructs new Adapter.
+     * @param _l2ArbitrumInbox Inbox helper contract to send messages to Arbitrum.
+     * @param _l2ERC20GatewayRouter ERC20 gateway router contract to send tokens to Arbitrum.
+     * @param _l3RefundL3Address L3 address to receive gas refunds on after a message is relayed.
+     * @param _l2Usdc Native USDC address on L2.
+     * @param _cctpTokenMessenger TokenMessenger contract to bridge via CCTP.
+     * @param _customGasTokenFunder Contract that funds the custom gas token.
+     * @param _l3MaxSubmissionCost Amount of gas token allocated to pay for the base submission fee. The base
+     * submission fee is a parameter unique to Arbitrum retryable transactions. This value is hardcoded
+     * and used for all messages sent by this adapter.
+     */
+    constructor(
+        ArbitrumInboxLike _l2ArbitrumInbox,
+        ArbitrumERC20GatewayLike _l2ERC20GatewayRouter,
+        address _l3RefundL3Address,
+        IERC20 _l2Usdc,
+        ICCTPTokenMessenger _cctpTokenMessenger,
+        FunderInterface _customGasTokenFunder,
+        uint256 _l3MaxSubmissionCost,
+        address _l3SpokePool,
+        address _crossDomainAdmin
+    )
+        CircleCCTPAdapter(_l2Usdc, _cctpTokenMessenger, CircleDomainIds.UNINITIALIZED)
```

Comment:
This is speculation on how CCTP would work on L3s. We may just want to remove this on the forwarders.